### PR TITLE
RPC get_balance implementation

### LINKED ISSUE #14: implement rpc get_balance
Restructure the current trusted getter get_balance such that the balance will not be called from stf state but from the balance hash map.

* authenticate sender and verify permissions (main, proxy)
* return tuple (free, reserved)

depends on #2, #3

### PR DESCRIPTION
**Pull request by murerfel**
Implemented the RPC API to get_balance:

1. verify signature
2. verify matching operation
3. verify proxy account (if provided) is authorized to access main account
4. get the balance from storage

Also did quite a bit of refactoring to de-couple the RPC API from the actual implementation in the polkadex module. Use trait objects to achieve that. When testing, mock implementation of those trait are used.

Resolves #14

### CODE REVIEW

File: enclave/src/polkadex_gateway.rs [hunk 1]
```diff
@@ -40,13 +43,25 @@ pub enum GatewayError {
...
 }
 
+impl alloc::fmt::Display for GatewayError {
+    fn fmt(&self, f: &mut alloc::fmt::Formatter) -> FormatResult {
+        write!(f, "{:?}", self)
+        // or, alternatively:
+        // fmt::Debug::fmt(self, f)
+    }
+}
+
```
Comment (murerfel) [inline enclave/src/polkadex_gateway.rs:53]: Added a Display implementation to allow converting the GatewayError to string

File: enclave/src/polkadex_gateway.rs [hunk 2]
```diff
@@ -40,13 +43,25 @@ pub enum GatewayError {
...
 /// 5. report OpenFinex API result to sender
-pub fn place_order(main_account: AccountId, proxy_acc: Option<AccountId>, order: Order) -> Result<OrderUUID, GatewayError> {
+pub fn place_order(
+    main_account: AccountId,
+    proxy_acc: Option<AccountId>,
+    order: Order,
+) -> Result<OrderUUID, GatewayError> {
     // Authentication
     authenticate_user(main_account.clone(), proxy_acc)?;
     // Mutate Balances
```
Comment (murerfel) [inline enclave/src/polkadex_gateway.rs:67]: The rest of the changes in this file came from cargo fmt, not real changes

File: enclave/src/rpc/polkadex_rpc_gateway.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+/// Gateway trait from RPC API -> Polkadex gateway implementation
+pub trait RpcGateway: Send + Sync {
+    fn authorize_user(
+        &self,
+        main_account: AccountId,
+        proxy_account: Option<AccountId>,
+    ) -> Result<(), GatewayError>;
+
+    fn get_balances(&self, main_account: AccountId, asset_it: AssetId) -> SgxResult<Balances>;
+}
```
Comment (murerfel) [inline enclave/src/rpc/polkadex_rpc_gateway.rs:33]: New Gateway Trait, is the connecting facade between the RPC API and the polkadex implementations. This trait can be mocked for unit tests.

File: enclave/src/rpc/rpc_get_balance.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+use jsonrpc_core::{BoxFuture, Params, Result as RpcResult, RpcMethodSync, Value};
+use log::*;
+use polkadex_sgx_primitives::types::DirectRequest;
+use substratee_stf::{Getter, TrustedGetter, TrustedOperation};
+use substratee_worker_primitives::DirectRequestStatus;
+
+pub struct RpcGetBalance {
+    top_extractor: Box<dyn TrustedOperationExtractor + 'static>,
+    rpc_gateway: Box<dyn RpcGateway + 'static>,
+}
```
Comment (murerfel) [inline enclave/src/rpc/rpc_get_balance.rs:36]: Use trait objects inside RpcGetBalance to promote de-coupling and allow clean unit testing

File: enclave/src/rpc/rpc_get_balance.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    fn method_impl(
+        &self,
+        request: DirectRequest,
+    ) -> Result<(Balances, bool, DirectRequestStatus), String> {
+        debug!("entering get_balance RPC");
+
+        let verified_trusted_operation =
+            self.top_extractor.get_verified_trusted_operation(request)?;
```
Comment (murerfel) [inline enclave/src/rpc/rpc_get_balance.rs:56]: Verify signature and decode the trusted operation

File: enclave/src/rpc/rpc_get_balance.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        let main_account = trusted_getter_signed.getter.main_account().clone();
+        let proxy_account = trusted_getter_signed.getter.proxy_account().clone();
+
+        let _authorization_result = match self
+            .rpc_gateway
+            .authorize_user(main_account.clone(), proxy_account)
+        {
+            Ok(()) => Ok(()),
+            Err(e) => Err(format!("Authorization error: {}", e)),
+        }?;
```
Comment (murerfel) [inline enclave/src/rpc/rpc_get_balance.rs:75]: Authorize using main account and optional proxy account

File: enclave/src/rpc/rpc_get_balance.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+        let asset_id = match trusted_getter_signed.getter {
+            TrustedGetter::get_balance(_, c, _) => Ok(c),
+            _ => Err(RpcCallStatus::operation_type_mismatch.to_string()),
+        }?;
```
Comment (murerfel) [inline enclave/src/rpc/rpc_get_balance.rs:80]: Gets the currency id, and implicitly makes sure we have the correct trusted operation

File: enclave/src/rpc/rpc_get_balance.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        });
+
+        let request = create_dummy_request();
+        let rpc_get_balance = RpcGetBalance::new(top_extractor, rpc_gateway);
+
+        let result = rpc_get_balance.method_impl(request).unwrap();
+        assert_eq!(result.2, DirectRequestStatus::Ok);
+        assert_eq!(result.0.free, free_balance);
+        assert_eq!(result.0.reserved, reserved_balance);
+    }
```
Comment (murerfel) [inline enclave/src/rpc/rpc_get_balance.rs:144]: Unit test to ensure we get the correct balance back

**Review by haerdib — APPROVED** [review]
LGTM

File: enclave/src/polkadex_gateway.rs (hunk 1 above)
Comment (Gauthamastro) [reply enclave/src/polkadex_gateway.rs:53]: This is good.

**Review by Gauthamastro — APPROVED** [review]

**Pull request merged**